lib/grant-types/refresh-token-grant-type.js

```diff
@@ -0,0 +1,175 @@
+  }
+
+  if (!options.model.revokeToken) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `revokeToken()`');
+  }
+
+  if (!options.model.saveToken) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `saveToken()`');
+  }
+
+  AbstractGrantType.call(this, options);
+}
+
+/**
+ * Inherit prototype.
+ */
+
+util.inherits(RefreshTokenGrantType, AbstractGrantType);
+
+/**
+ * Handle refresh token grant.
+ *
+ * @see https://tools.ietf.org/html/rfc6749#section-6
+ */
+
+RefreshTokenGrantType.prototype.handle = function(request, client) {
+  if (!request) {
+    throw new InvalidArgumentError('Missing parameter: `request`');
+  }
+
+  if (!client) {
+    throw new InvalidArgumentError('Missing parameter: `client`');
+  }
+
+  var scope = this.getScope(request);
+
+  return Promise.bind(this)
+    .then(function() {
+      return this.getRefreshToken(request, client);
+    })
+    .tap(function(token) {
+      return this.revokeToken(token);
+    })
+    .then(function(token) {
+      return this.saveToken(token.user, client, scope);
+    });
+};
+
+/**
+ * Get refresh token.
+ */
+
+RefreshTokenGrantType.prototype.getRefreshToken = function(request, client) {
+  if (!request.body.refresh_token) {
+    throw new InvalidRequestError('Missing parameter: `refresh_token`');
+  }
+
+  if (!is.vschar(request.body.refresh_token)) {
+    throw new InvalidRequestError('Invalid parameter: `refresh_token`');
+  }
+
+  return Promise.try(this.model.getRefreshToken, request.body.refresh_token)
+    .then(function(token) {
+      if (!token) {
+        throw new InvalidGrantError('Invalid grant: refresh token is invalid');
+      }
+
+      if (!token.client) {
+        throw new ServerError('Server error: `getRefreshToken()` did not return a `client` object');
+      }
+
+      if (!token.user) {
+        throw new ServerError('Server error: `getRefreshToken()` did not return a `user` object');
+      }
+
+      if (token.client.id !== client.id) {
+        throw new InvalidGrantError('Invalid grant: refresh token is invalid');
+      }
+
+      if (token.refreshTokenExpiresAt && !(token.refreshTokenExpiresAt instanceof Date)) {
+        throw new ServerError('Server error: `refreshTokenExpiresAt` must be a Date instance');
+      }
+
+      if (token.refreshTokenExpiresAt && token.refreshTokenExpiresAt < new Date()) {
+        throw new InvalidGrantError('Invalid grant: refresh token has expired');
+      }
+
+      return token;
+    });
+};
+
+/**
+ * Revoke the refresh token.
+ *
+ * @see https://tools.ietf.org/html/rfc6749#section-6
+ */
+
+RefreshTokenGrantType.prototype.revokeToken = function(token) {
+  return Promise.try(this.model.revokeToken, token)
+    .then(function(token) {
+      if (!token) {
+        throw new InvalidGrantError('Invalid grant: refresh token is invalid');
+      }
+
+      if (!(token.refreshTokenExpiresAt instanceof Date)) {
```

Comment:
Ah, I think I see the confusion. Expiration was the method I chose for the model to indicate that a token has been revoked.

This could also be done with a boolean, for example: `isExpired: true` but since we already need to support expiration why not use the same mechanism?